[jiaaaaen] iP

jiaaaaen opened the pull request:
> “Never mistake motion for action.” – Ernest Hemingway [(source)](https://dansilvestre.com/productivity-quotes/)
# Bwead will help u track your tasks! :rofl:
Bwead can track:
* todo tasks
* deadline tasks
* event tasks

**FOR FREEEE**
to download Bwead,
1. download the .jar file from [here](https://github.com/jiaaaaen/ip/releases/tag/A-Jar)
2. open your terminal
3. run the command `java -jar "{filename}.jar"`
4. enjoy Bwead!

the main features are: 
- [ ] listing all tasks
- [x] adding and deleting tasks
- [ ] marking as done/ undone
- [ ] tasks saved locally

if you are a Java programmer, you can use it to practice Java too. here's the `main` method:
```ruby
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

jiaaaaen committed:
rename, greet, exit

jiaaaaen committed:
echo

jiaaaaen committed:
add, list

jiaaaaen committed:
mark as done

jiaaaaen committed:
todo, event, deadline

jiaaaaen committed:
automated text ui testing

jiaaaaen committed:
handle errors

jiaaaaen committed:
delete

jiaaaaen committed:
fixed error handling

jiaaaaen committed:
save tasks in a file

jiaaaaen committed:
understand dates

jiaaaaen committed:
implement oop

jiaaaaen committed:
Merge branch 'A-MoreOOP'

* A-MoreOOP:
  implement oop

jiaaaaen committed:
put all classes in one package

jiaaaaen committed:
Merge branch 'A-Packages'

* A-Packages:
  put all classes in one package

jiaaaaen committed:
Merge commit '81a9c537afe8492675eb253968939364d65db091'

* commit '81a9c537afe8492675eb253968939364d65db091':
  build.gradle: Prevent generating a second JAR file
  Bump gradle and lib version
  Add Gradle support

jiaaaaen committed:
Automate project builds using Gradle

jiaaaaen committed:
Merge branch 'A-Gradle'

* A-Gradle:
  Automate project builds using Gradle

jiaaaaen committed:
add times

jiaaaaen committed:
Add JUnit tests

jiaaaaen committed:
Merge branch 'A-JUnit'

* A-JUnit:
  Add JUnit tests

jiaaaaen committed:
fix loading error message

jiaaaaen committed:
Add JavaDoc comments

jiaaaaen committed:
Tweak the code to comply with a coding standard

jiaaaaen committed:
Give users a way to find a task by searching for a keyword.

jiaaaaen committed:
Merge tag 'A-JavaDoc'

* tag 'A-JavaDoc':
  Add JavaDoc comments

jiaaaaen committed:
resolve merge conflict

jiaaaaen committed:
Merge branch 'Level-9'

* Level-9:
  Give users a way to find a task by searching for a keyword.

woke02 commented on src/main/java/Bwead/Ui.java:
```diff
@@ -0,0 +1,142 @@
+                    history.updateFile(taskList.getCurrentList());
+                    System.out.println("Got it. I've added this task: " + task.toString());
+                    System.out.println("Now you have " + taskList.getCurrentList().size() + " tasks in the list.");
+                } else if (input.startsWith("delete")) {
+                    Task toremove = taskList.getCurrentList().get(parser.getDeleteIndex() - 1);
+                    taskList.getCurrentList().remove(toremove);
+                    history.updateFile(taskList.getCurrentList());
+                    System.out.println("Noted. I've removed this task: " + toremove.toString());
+                    System.out.println("Now you have " + taskList.getCurrentList().size() + " tasks in the list.");
```
Can consider handling commands using a Parser/ CommandHandler class with the different commands as classes

woke02 reviewed (commented):
Overall, code follows coding standard, OOP is applied appropriately. Some of the classes are getting quite long (like Ui class). Consider splitting into smaller classes.

jiaaaaen committed:
Use checkStyle to detect coding style violations.

jiaaaaen committed:
Merge tag 'A-CheckStyle'

* tag 'A-CheckStyle':
  Use checkStyle to detect coding style violations.

choiwab commented on src/main/java/Bwead/Bwead.java:
```diff
@@ -0,0 +1,60 @@
+ */
+public class Bwead {
+
+    private static String name = "Bwead";
+    private History history;
+    private TaskList tasks;
+    private Ui ui;
```
I like how intuitive the variable names are. Good job!

choiwab commented on src/main/java/Bwead/Bwead.java:
```diff
@@ -0,0 +1,60 @@
+    private Ui ui;
+
+    /**
+     * Constructs a Bwead instance using a file path which points to local file.
+     * a new ui, and a history is created with the file path.
+     * a taskList will be instantiated with tasks from the saved file, if any. else, an empty one will be created.
+     * if load() throws BweadException due to file not existing, the taskList will be empty.
+     *
+     * @param filePath file path of the local file which will save tasks.
+     * @throws IllegalArgumentException IOException when a failure occurs while performing read or write
+     *     operations from the load() method.
+     */
```
The JavaDoc for this method is very specific and direct. Very good job!

choiwab commented on src/main/java/Bwead/History.java:
```diff
@@ -0,0 +1,121 @@
+            text1 = text1.substring(0, text1.length() - 2);
+            String dates = text.split("from: ")[1];
+            String startString = dates.split(" to: ")[0];
+            String startDateString = startString.substring(0, startString.length() - 7);
+            String startTimeString = startString.substring(startString.length() - 5, startString.length() - 3)
+                    + startString.substring(startString.length() - 2, startString.length());
+            String endString = dates.split(" to: ")[1];
+            endString = endString.substring(0, endString.length() - 1);
+            String endDateString = endString.substring(0, endString.length() - 7);
+            String endTimeString = endString.substring(endString.length() - 5, endString.length() - 3)
+                    + endString.substring(endString.length() - 2, endString.length());
+            LocalTime startTime = LocalTime.parse(startTimeString.substring(0, 2) + ":"
+                    + startTimeString.substring(2, 4));
+            LocalTime endTime = LocalTime.parse(endTimeString.substring(0, 2) + ":" + endTimeString.substring(2, 4));
+            return new Event(text1 + " ", LocalDate.parse(startDateString, DateTimeFormatter.ofPattern("MMM d uuuu")),
+                    LocalDate.parse(endDateString, DateTimeFormatter.ofPattern("MMM d uuuu")),
+                    startTime, endTime);
```
Very good job not making the common mistake of not spacing between arithmetic equations. Big ups.

choiwab commented on src/main/java/Bwead/Parser.java:
```diff
@@ -0,0 +1,197 @@
+     *
+     * @return boolean true if the command is invalid and multiple words.
+     */
+    public boolean isInvalidMultiWord() {
+        return !command.startsWith("todo") && !command.startsWith("deadline") && !command.startsWith("event")
+                && !command.startsWith("delete") && !command.startsWith("mark") && !command.startsWith("unmark")
+                && !command.startsWith("find");
+    }
```
Again, good job with spacing here. Perfect.

choiwab commented on src/main/java/Bwead/History.java:
```diff
@@ -0,0 +1,121 @@
+package Bwead;
+
+import java.io.File;
+import java.io.FileWriter;
+import java.io.IOException;
+import java.time.LocalDate;
+import java.time.LocalTime;
+import java.time.format.DateTimeFormatter;
+import java.util.ArrayList;
+import java.util.Scanner;
```
I like how you specified every imported class and did not use "*"

choiwab reviewed (commented):
I don't see any nits. Good job keeping the Java Coding Standard. LGTM overall!

jiaaaaen committed:
Add a GUI to the chatbot. Use the JavaFX technology to implement the GUI.

jiaaaaen committed:
Merge branch 'Level-10'

* Level-10:
  Add a GUI to the chatbot. Use the JavaFX technology to implement the GUI.

thisisaditya17 commented on src/main/java/Bwead/Deadline.java:
```diff
@@ -0,0 +1,55 @@
+        } else {
+            str = " ";
+        }
+        return "[D][" + str + "] " + text + " (by: " + date.format(DateTimeFormatter.ofPattern("MMM d yyyy")) + ", "
```
You can consider creating a static formatter of the date-time output in the Task class. Just a better OOP practice 😄

thisisaditya17 commented on src/main/java/Bwead/Task.java:
```diff
@@ -0,0 +1,47 @@
+        this.isDone = isDone;
+    }
+
+    public String getName() {
```
I feel this is very vague.  🫤 
Suggestion: can change it something like getDescription() because the description doesn't contain just the name.
Also this is repetitive as you already have a getText() method doing the same thing. 🤔

thisisaditya17 commented on src/main/java/Bwead/TaskList.java:
```diff
@@ -0,0 +1,50 @@
+
+    private static ArrayList<Task> currentList;
+
+    /*
```
Why is it in comments? 🤔 
If not useful, can just delete it. 
It will make the code base look cleaner.

thisisaditya17 commented on src/test/java/Bwead/BweadTest.java:
```diff
@@ -0,0 +1,107 @@
+package Bwead;
```
this package seems too packed rn 🤔 

Suggestion: Can separate files according to their uses into different directories of the package like Bwead.main, Bwead.tasks, Bwead.tests, etc.

thisisaditya17 commented on src/main/java/Bwead/History.java:
```diff
@@ -0,0 +1,121 @@
+            String text = string.substring(7);
+            return new Todo(text);
+        } else if (type == 'D') {
+            String text = string.substring(7);
```
Can consider creating a method to process the data read from the save file to process and separate everything in the Parser class. 🕺🏼

thisisaditya17 reviewed (commented):
Overall, a very solid code.
I really like how you have different classes handling different tasks following OOP to the core. 😆 
Can consider my suggestions. 🕺🏼

jiaaaaen committed:
use assert features

assert is used to make sure the task names cannot be null in the getName methods.

assert is used to make sure there is a history and task list object made for Bwead.

assert is used to make sure the task list contains more than one tasks when the delete command is used.

assert is used to make sure the task list is not empty, especially after adding a new task to it.

jiaaaaen committed:
Improve code quality

Shortened a very long method in the ui class by splitting into smaller methods.

jiaaaaen committed:
Merge pull request #2 from jiaaaaen/A-Assertions

use assert features

jiaaaaen committed:
Merge branch 'master' of https://github.com/jiaaaaen/ip

* 'master' of https://github.com/jiaaaaen/ip:
  use assert features

jiaaaaen committed:
Merge branch 'master' into A-CodeQuality

jiaaaaen committed:
Merge branch 'master' into A-CodeQuality

* master:
  use assert features

# Conflicts:
#	src/main/java/Bwead/Ui.java

jiaaaaen committed:
Merge pull request #3 from jiaaaaen/A-CodeQuality

Improve code quality

jiaaaaen committed:
Merge branch 'A-CodeQuality' of https://github.com/jiaaaaen/ip into A-CodeQuality

* 'A-CodeQuality' of https://github.com/jiaaaaen/ip:

# Conflicts:
#	src/main/java/Bwead/Ui.java

jiaaaaen committed:
Merge tag 'A-CodeQuality'

* tag 'A-CodeQuality':
  Improve code quality

# Conflicts:
#	src/main/java/Bwead/Ui.java

jiaaaaen committed:
Merge branch 'master' of https://github.com/jiaaaaen/ip

* 'master' of https://github.com/jiaaaaen/ip:

jiaaaaen committed:
add a snooze extension

jiaaaaen committed:
Merge tag 'B-Snooze'

* tag 'B-Snooze':
  add a snooze extension

jiaaaaen committed:
Update README.md

jiaaaaen committed:
Add files via upload

jiaaaaen committed:
Update README.md

jiaaaaen committed:
Update README.md

jiaaaaen committed:
Update README.md

jiaaaaen committed:
Update README.md

jiaaaaen committed:
Update README.md

jiaaaaen committed:
Update README.md

jiaaaaen committed:
edit time format in strings

jiaaaaen committed:
Merge branch 'master' of https://github.com/jiaaaaen/ip

* 'master' of https://github.com/jiaaaaen/ip:
  Update README.md
  Update README.md
  Update README.md
  Update README.md
  Update README.md
  Update README.md
  Add files via upload
  Update README.md

jiaaaaen committed:
fix file opening for new computers

jiaaaaen committed:
Update README.md

jiaaaaen committed:
Update README.md

jiaaaaen committed:
Update README.md

jiaaaaen committed:
update junit tests

jiaaaaen committed:
Update README.md

jiaaaaen committed:
Update README.md

jiaaaaen committed:
finalise error handling

jiaaaaen committed:
Merge branch 'master' into A-CodeQuality

jiaaaaen committed:
Merge pull request #4 from jiaaaaen/A-CodeQuality

Improve code quality

jiaaaaen committed:
add more javadoc for event and dealine
